File: backend/src/templates.ts
import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { TEMPLATES, TemplateDef } from "./template-defs.js";

/**
 * Template store backing the `list_templates` / `load_template` agent tools.
 * The placeholder inventory goes to the model; the .pptx base64 never does —
 * it travels only in the relayed `insert_template_slides` payload to the pane.
 */

// Works from both src/ (tsx) and dist/ (tsc build) — templates/ is a sibling.
const TEMPLATES_DIR = path.join(path.dirname(fileURLToPath(import.meta.url)), "..", "templates");

export class TemplateToolError extends Error {}

function requireTemplate(templateId: string): TemplateDef {
  const def = TEMPLATES.find((t) => t.id === templateId);
  if (!def) {
    const ids = TEMPLATES.map((t) => t.id).join(", ");
    throw new TemplateToolError(`unknown template_id "${templateId}" — available templates: ${ids}`);
  }
  return def;
}

export function listTemplates() {
  return {
    templates: TEMPLATES.map((t) => ({
      id: t.id,
      name: t.name,
      description: t.description,
      slideCount: t.slides.length,
    })),
  };
}

/** Full placeholder inventory the model plans `apply_slide_content` fills against. */
export function loadTemplate(templateId: string) {
  const { id, name, description, slides } = requireTemplate(templateId);
  return { id, name, description, slides };
}

/** The .pptx itself — relayed to the task pane for insertSlidesFromBase64. */
export function getTemplateBase64(templateId: string): string {
  const def = requireTemplate(templateId);
  try {
    return readFileSync(path.join(TEMPLATES_DIR, def.file)).toString("base64");
  } catch {
    throw new TemplateToolError(
      `template file "${def.file}" is missing — run \`npm run build:templates\` in backend/`,
    );
  }
}

File: backend/app/templates.py
"""Template store backing the `list_templates` / `load_template` agent tools —
the Python port of backend/src/templates.ts.

Definitions come from the repo-root `templates.json` (the single source of truth
shared with the Node generator). The placeholder inventory goes to the model; the
.pptx base64 never does — it travels only in the relayed `insert_template_slides`
payload to the pane.
"""

from __future__ import annotations

import base64
import json
from functools import lru_cache
from pathlib import Path
from typing import Any

# backend-py/app/templates.py -> repo root is two parents up.
_REPO_ROOT = Path(__file__).resolve().parent.parent.parent
_TEMPLATES_JSON = _REPO_ROOT / "templates.json"
# The committed .pptx artifacts live alongside this package.
_TEMPLATES_DIR = Path(__file__).resolve().parent.parent / "templates"


class TemplateToolError(Exception):
    """Recoverable tool error surfaced back to the model as an `Error: ...` string."""


@lru_cache(maxsize=1)
def _load_defs() -> list[dict[str, Any]]:
    return json.loads(_TEMPLATES_JSON.read_text(encoding="utf-8"))


def _require_template(template_id: str) -> dict[str, Any]:
    defs = _load_defs()
    for t in defs:
        if t["id"] == template_id:
            return t
    ids = ", ".join(t["id"] for t in defs)
    raise TemplateToolError(f'unknown template_id "{template_id}" — available templates: {ids}')


def list_templates() -> dict[str, Any]:
    return {
        "templates": [
            {
                "id": t["id"],
                "name": t["name"],
                "description": t["description"],
                "slideCount": len(t["slides"]),
            }
            for t in _load_defs()
        ]
    }


def load_template(template_id: str) -> dict[str, Any]:
    """Full placeholder inventory the model plans `apply_slide_content` fills against."""
    t = _require_template(template_id)
    return {"id": t["id"], "name": t["name"], "description": t["description"], "slides": t["slides"]}


def get_template_base64(template_id: str) -> str:
    """The .pptx itself — relayed to the task pane for insertSlidesFromBase64."""
    t = _require_template(template_id)
    path = _TEMPLATES_DIR / t["file"]
    try:
        return base64.b64encode(path.read_bytes()).decode("ascii")
    except OSError as err:
        raise TemplateToolError(
            f'template file "{t["file"]}" is missing — run '
            "`python scripts/build_templates.py` in backend/"
        ) from err
